src/dialog/unix/SDL_zenitydialog.c

```diff
@@ -80,184 +59,167 @@ char *zenity_clean_name(const char *name)
+ *     zenity --file-selection --separator=\n [--multiple]
  *                         [--directory] [--save --confirm-overwrite]
  *                         [--filename FILENAME] [--modal --attach 0x11w1nd0w]
  *                         [--title TITLE] [--ok-label ACCEPT]
  *                         [--cancel-label CANCEL]
  *                         [--file-filter=Filter Name | *.filt *.fn ...]...
  */
-static char** generate_args(const zenityArgs* info)
+static bool generate_zenity_args(SDL_FileDialogType type, SDL_DialogFileCallback callback, void *userdata, SDL_PropertiesID props, zenityArgs *args)
 {
-    int argc = 4;
-    int nextarg = 0;
-    char **argv = NULL;
-
-    // ARGC PASS
-    if (info->allow_many) {
-        argc++;
-    }
-
-    switch (info->type) {
-    case SDL_FILEDIALOG_OPENFILE:
-        break;
-
-    case SDL_FILEDIALOG_SAVEFILE:
-        argc += 2;
-        break;
-
-    case SDL_FILEDIALOG_OPENFOLDER:
-        argc++;
-        break;
-    };
-
-    if (info->filename) {
-        argc += 2;
-    }
-
-    if (info->x11_window_handle) {
-        argc += 3;
-    }
-
-    if (info->title) {
-        argc += 2;
-    }
-
-    if (info->accept) {
-        argc += 2;
-    }
+    SDL_zerop(args);
+    args->callback = callback;
+    args->userdata = userdata;
+    args->nfilters = SDL_GetNumberProperty(props, SDL_PROP_FILE_DIALOG_NFILTERS_NUMBER, 0);
 
-    if (info->cancel) {
-        argc += 2;
+    const char **argv = SDL_malloc(
+        sizeof(*argv) * (3   /* zenity --file-selection --separator=\n */
+                         + 1 /* --multiple */
+                         + 1 /* --directory */
+                         + 2 /* --save --confirm-overwrite */
+                         + 2 /* --filename [file] */
+                         + 3 /* --modal --attach [handle] */
+                         + 2 /* --title [title] */
+                         + 2 /* --ok-label [label] */
+                         + 2 /* --cancel-label [label] */
+                         + args->nfilters + 1 /* NULL */));
+    if (!argv) {
+        return false;
     }
+    args->argv = argv;
 
-    if (info->filters) {
-        argc += info->nfilters;
+    /* Properties can be destroyed as soon as the function returns; copy over what we need. */
+#define COPY_STRING_PROPERTY(dst, prop)                             \
+    {                                                               \
+        const char *str = SDL_GetStringProperty(props, prop, NULL); \
+        if (str) {                                                  \
+            dst = SDL_strdup(str);                                  \
+            if (!dst) {                                             \
+                goto cleanup;                                       \
+            }                                                       \
+        }                                                           \
     }
```

Comment:
@slouken Do macros need to follow the `do { ... } while(0)` pattern?

Reply:
Yes, in general.